modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java

```diff
@@ -2418,16 +2418,35 @@ public CodegenModel fromModel(String name, Schema schema) {
         m.classFilename = toModelFilename(name);
         m.modelJson = Json.pretty(schema);
         m.externalDocumentation = schema.getExternalDocs();
-        if (schema.getExtensions() != null && !schema.getExtensions().isEmpty()) {
-            m.getVendorExtensions().putAll(schema.getExtensions());
-        }
         m.isAlias = (typeAliases.containsKey(name)
                 || isAliasOfSimpleTypes(schema)); // check if the unaliased schema is an alias of simple OAS types
         m.setDiscriminator(createDiscriminator(name, schema, this.openAPI));
         if (!this.getLegacyDiscriminatorBehavior()) {
             m.addDiscriminatorMappedModelsImports();
         }
 
+        if (schema.getExtensions() != null && !schema.getExtensions().isEmpty()) {
+            m.getVendorExtensions().putAll(schema.getExtensions());
+            if (ModelUtils.isComposedSchema(schema)
+                    && m.discriminator == null
+                    && schema.getExtensions().containsKey("x-one-of-name")) {
+                boolean isDeductionCase = true;
+                List<String> deductionModelNames = new ArrayList<>();
+                for (Schema modelSchema : ((ComposedSchema) schema).getOneOf()) {
+                    if (modelSchema.get$ref() == null) {
+                        isDeductionCase = false;
+                        break;
+                    }
+                    String modelName = ModelUtils.getSimpleRef(modelSchema.get$ref());
+                    deductionModelNames.add(toModelName(modelName));
+                }
+                if (isDeductionCase) {
+                    m.vendorExtensions.put("x-deduction", true);
+                    m.vendorExtensions.put("x-deduction-model-names", deductionModelNames);
+                }
+            }
+        }
+
```

Comment:
FYI. Since this changes the default codegen, it will take a lot more time to review.